### ISSUE TITLE
Support for typescript and .tsx

### ISSUE THREAD

**Issue by amirhouieh** [brandonweiss/charge#666]
Hi there, just came across Charge. Reading the idea and concept behind it sounds great!
Just wondering how difficult is to implement support for `typescript` and `.tsx` ? I saw that Charge uses babel and so considering [TS support in Babel 7](https://babeljs.io/docs/en/babel-preset-typescript.html) would not it be an easy implementation? I could also make PR if you could point out where I should look into!

**Comment by brandonweiss** [comment]
Hey, thanks!

I’ve only used TypeScript a little bit, but I’m definitely interested in it. I’ve actually been considering converting this codebase to TypeScript—I think it might benefit from it.

As for allowing TypeScript in “source” files… that’s interesting 🤔 Offhand, here are my thoughts.

In Charge there are two types of “source” files where TypeScript could potentially be used: pages (`.html.jsx`) and client-side JavaScript (`.js`). As you mentioned, given Babel supports TypeScript the actual technical work involved in supporting TypeScript is rather small, however it’s not really clear to me that it would be particularly beneficial to most people.

My limited experience with TypeScript tells me that it’s most useful for large and/or complex applications. It’s not that you can’t use it for something small—you can—but it probably wouldn’t make sense or be worth the effort. I think most websites using Charge are quite small and simple. The design of Charge is sort of a natural constraint on complexity. It’s the tradeoff I made for zero configuration. For example, all my small sites use Charge, and I even converted my blog to it, but after a little while I converted my blog to Gatsby, because there were some more complex things I wanted to do with it that wouldn’t have been possible with Charge.

There’s also one other potentially bigger problem: the whole point of Charge is that it’s zero-config. TypeScript has a _lot_ of configuration options. Would it be possible to choose one configuration that “just worked” for everyone? My gut says probably not. I think there would be much debate over what `noImplicitAny` any (and every other option) should be set to.

Let me know what you think and if I missed any pros/cons!

**Comment by amirhouieh** [comment]
Hey, thanks for the explanation. 
Well I do understand your points over simplicity and zero-config over Charge. However I do not think that using TS necessary would make it complex. I think we are talking about giving an option besides what is already offering. So who ever pick TS then have probably knows why and so this would not effect. 
For example I have made a promise to myself to now touch JS (you might guess why) since 3 years ago, and ever since I have succeed to keep that promise by using TS only. 
Regarding zero-config, no there is not one single config which would “just worked” for everyone! However it is not much of configuration it's only one file `tsconfig.json` which again whoever choose TS over JS has already know this too and nothing extra for them. 

So my idea here would be to enable TS option for both type sources `html.tsx` and client scripts `ts`. however the second one is not very important and could still be out of the loop and leave it for users to decide!

**Comment by brandonweiss** [comment]
Thanks for your thoughts!

I wasn’t necessarily saying that using TypeScript would make someone’s site more complex, I was saying that small sites might not benefit much from being written in TypeScript, and so supporting TypeScript in Charge might not be worth the effort and added complexity given the amount of usage that feature might get.

You mentioned that TypeScript configuration is “not much of one”. I really disagree here. It might only be one _file_, but that file has something like ~90 configuration options. But to be clear, it wouldn’t really matter if TypeScript had exactly one configuration option—Charge is zero-config (not one-config). I can imagine that limitation might seem annoying or inconvenient, but that limitation is part of what (I think) makes Charge so great, or at least so unique. If I broke the zero-config rule, Charge wouldn’t really have anything that differentiates it from other static site generators.

To put it another way, there is no one perfect tool that can solve any/every problem. If what you value is a static site generator that’s really simple, has no configuration, and will allow you to build a small site really quickly, then Charge might be for you. But if what you value is using TypeScript because you promised yourself never to write plain JavaScript again (which I understand), then there are plenty of other static site generators out there that support TypeScript, like [Gatsby](https://www.gatsbyjs.org), [Stasis](https://getstasis.com), or [Hydrogen](https://hydrogenjs.org), although the tradeoff is that they’re much more complex to use.

I’ll leave this issue open for a while to collect if anyone else is interested in using TypeScript with Charge!
